File: Solutions/Python/pow_x_n.py
"""
    [MEDIUM]
    50. Pow(x, n)

    Concepts:
    - math
    - recursion

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 16.81 MB  [Beats 100%]
"""

class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1.0
        elif n < 0:
            return 1 / self.myPow(x, -n)
        elif n % 2:
            return self.myPow(x * x, n // 2) * x
        else:
            return self.myPow(x * x, n // 2)

File: Solutions/Cpp/pow_x_n.cpp
/*
    [MEDIUM]
    50. Pow(x, n)

    Concepts:
    - math
    - recursion

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 8.48 MB   [Beats 92.41%]
*/

class Solution {
public:
    double myPow(double x, long n) {
        if (n == 0)
            return 1.0;
        else if (n < 0)
            return 1.0 / myPow(x, -n);
        else if (n & 1)
            return myPow(x * x, n / 2) * x;
        else
            return myPow(x * x, n / 2);
    }
};